cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java

```diff
@@ -0,0 +1,239 @@
+      @Nullable SourceControlMeta sourceControlMeta)
+      throws IOException {
+    StructuredTable scmTable = getNamespaceSourceControlMetadataTable();
+    scmTable.upsert(getNamespaceSourceControlMetaFields(appId, sourceControlMeta));
+  }
+
+  /**
+   * Retrieves the source control metadata for the specified application ID in the namespace from
+   * {@code NamespaceSourceControlMetadata} table.
+   *
+   * @param appId {@link ApplicationId} for which the source control metadata is being retrieved.
+   * @return The {@link SourceControlMeta} associated with the application ID, or {@code null} if no
+   *         metadata is found.
+   * @throws IOException If it fails to read the metadata.
+   */
+  @Nullable
+  public SourceControlMeta getNamespaceSourceControlMeta(ApplicationId appId) throws IOException {
+    List<Field<?>> primaryKey = getPrimaryKey(
+        appId.getNamespace(), appId.getEntityType().toString(), appId.getEntityName());
+    StructuredTable table = getNamespaceSourceControlMetadataTable();
+    Optional<StructuredRow> row = table.read(primaryKey);
+    if (row.isPresent()) {
+      StructuredRow nonNullRow = row.get();
+      String specificationHash = nonNullRow
+          .getString(StoreDefinition.NamespaceSourceControlMetadataStore.SPECIFICATION_HASH_FIELD);
+      String commitId = nonNullRow.getString(
+          StoreDefinition.NamespaceSourceControlMetadataStore.COMMIT_ID_FIELD);
+      Long lastSynced = nonNullRow.getLong(
+          StoreDefinition.NamespaceSourceControlMetadataStore.LAST_MODIFIED_FIELD);
+      if (specificationHash == null && commitId == null && lastSynced == 0) {
+        return null;
+      }
+      return new SourceControlMeta(specificationHash, commitId, Instant.ofEpochMilli(lastSynced));
+    }
+    return null;
+  }
+
+  private Collection<Field<?>> getNamespaceSourceControlMetaFields(ApplicationId appId,
+      SourceControlMeta scmMeta) {
+    List<Field<?>> fields = getPrimaryKey(appId);
+    fields.add(Fields.stringField(
+        StoreDefinition.NamespaceSourceControlMetadataStore.SPECIFICATION_HASH_FIELD,
+        scmMeta == null ? "" : scmMeta.getFileHash()));
+    fields.add(
+        Fields.stringField(StoreDefinition.NamespaceSourceControlMetadataStore.COMMIT_ID_FIELD,
+            scmMeta == null ? "" : scmMeta.getCommitId()));
+    fields.add(
+        Fields.longField(StoreDefinition.NamespaceSourceControlMetadataStore.LAST_MODIFIED_FIELD,
+            scmMeta == null ? 0L : scmMeta.getLastSyncedAt().toEpochMilli()));
+    fields.add(
+        Fields.booleanField(StoreDefinition.NamespaceSourceControlMetadataStore.IS_SYNCED_FIELD,
+            scmMeta == null ? false : true));
+    return fields;
+  }
+
+  private List<Field<?>> getPrimaryKey(ApplicationId appId) {
+    return getPrimaryKey(appId.getNamespace(), appId.getEntityType().toString(),
+        appId.getEntityName());
+  }
+
+  private List<Field<?>> getPrimaryKey(String namespace, String type, String name) {
```

Comment:
why is this function neede ? I can see we are always taking ApplicationID as input

Reply:
Right. I have removed it.